### PR TITLE
Create builders for Vulkan::API objects

### PR DESCRIPTION
**Pull request by Quentin01**

### CODE REVIEW

**Review by saveman71 — DISMISSED** [review]

File: include/lug/Graphics/Vulkan/API/Builder/Device.hpp [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+
+    Device(const Device&&) = delete;
+    Device(Device&&) = delete;
+
+    Device& operator=(const Device&&) = delete;
+    Device& operator=(Device&&) = delete;
+
+    ~Device() = default;
+
+    void setExtensions(std::vector<const char*> extensions);
```
Comment (saveman71) [inline include/lug/Graphics/Vulkan/API/Builder/Device.hpp]: Add `// Setters` comment

File: include/lug/Graphics/Vulkan/API/Builder/Device.hpp [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+
+class Device {
+private:
+    struct QueueCreateInfo {
+        int8_t queueFamilyIndex;
+        std::vector<float> queuePriorities;
+    };
+
+public:
+    Device(PhysicalDeviceInfo& physicalDeviceInfo);
```
Comment (saveman71) [inline include/lug/Graphics/Vulkan/API/Builder/Device.hpp]: `const&`

File: src/lug/Graphics/Vulkan/API/Builder/Device.cpp [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+
+void Device::addQueue(VkQueueFlags queueFlags, uint8_t queueCount) {
+    int8_t queueFamilyIndex;
+    bool queueFound;
+    // Search the queue family index corresponding to the queue flags
+    queueFound = _physicalDeviceInfo.containsQueueFlags(queueFlags, queueFamilyIndex);
+
+    // queueFlags contains VK_QUEUE_TRANSFER_BIT and queue is not found,
+    // we can use either VK_QUEUE_GRAPHICS_BIT or VK_QUEUE_COMPUTE_BIT instead of VK_QUEUE_TRANSFER_BIT
+    if (!queueFound && (queueFlags & VK_QUEUE_TRANSFER_BIT))
```
Comment (saveman71) [inline src/lug/Graphics/Vulkan/API/Builder/Device.cpp]: `) {`

File: src/lug/Graphics/Vulkan/API/Builder/Device.cpp [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+    {
+        // We try to find a queue first by replacing VK_QUEUE_TRANSFER_BIT by VK_QUEUE_TRANSFER_BIT
+        // then by VK_QUEUE_COMPUTE_BIT
+        VkQueueFlags queueFlagsTmp = queueFlags & ~VK_QUEUE_TRANSFER_BIT;
+        queueFound = _physicalDeviceInfo.containsQueueFlags(queueFlagsTmp | VK_QUEUE_GRAPHICS_BIT, queueFamilyIndex) ||
+            _physicalDeviceInfo.containsQueueFlags(queueFlagsTmp | VK_QUEUE_COMPUTE_BIT, queueFamilyIndex);
+    }
+
+    // Queue not found
+    if (!queueFound)
```
Comment (saveman71) [inline src/lug/Graphics/Vulkan/API/Builder/Device.cpp]: `) {`

**Review by saveman71 — CHANGES_REQUESTED** [review]

**Review by saveman71 — DISMISSED** [review]

**Review by Nokitoo — APPROVED** [review]

**Pull request merged**